### PR TITLE
Isolate Application object in Env:

### PR DESCRIPTION
**Pull request by vinniefalco**
This change causes each instance of Env to construct its own isolated Application object for testing. Also included is part of a framework to create multiple Application objects in the same unit test and connect them together.

Reviewers: @ximinez @scottschurr

### CODE REVIEW

**Comment by scottschurr** [comment]
A clang debug build gives me the following warning:
```
src/ripple/test/jtx/impl/Env.cpp:79:7: warning: field 'bundle_' will be
      initialized after field 'test' [-Wreorder]
    : bundle_ (makeConfig())
```

**Comment by vinniefalco** [comment]
@scottschurr Updated

**Comment by scottschurr** [comment]
I have to say, `app.rpc()` is very cool.  I'm looking forward to using this.

File: src/ripple/app/main/Main.cpp [hunk 1]
```diff
@@ -204,18 +193,10 @@ static int runShutdownTests (std::unique_ptr<Config> config)
...
+static int runUnitTests(
     std::string const& pattern,
     std::string const& argument)
 {
-    // Config needs to be set up before creating Application
-    setupConfigForUnitTests (*config);
-
-    auto app = make_Application (
-        std::move(config),
-        std::make_unique<Logs>());
```
Comment (ximinez) [inline src/ripple/app/main/Main.cpp]: Nice!

File: src/ripple/test/jtx/impl/Env.cpp [hunk 2]
```diff
@@ -50,17 +50,41 @@ namespace test {
 
 namespace jtx {
 
+Env::AppBundle::AppBundle(Application* app_)
+    : app (app_)
+{
+}
```
Comment (ximinez) [inline src/ripple/test/jtx/impl/Env.cpp]: This doesn't look like it's used anywhere. Is it "part of a framework to create multiple Application objects in the same unit test and connect them together" that you mention in the PR description?

File: src/ripple/test/mao/impl/Net_test.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    void
+    run() override
+    {
+        testStartStop();
+        testRPC();
+    }
+};
+
+BEAST_DEFINE_TESTSUITE_MANUAL(Net,mao,ripple)
```
Comment (ximinez) [inline src/ripple/test/mao/impl/Net_test.cpp]: What does `mao` mean in this context?

**Comment by vinniefalco** [comment]
@scottschurr Thanks! The intent is to allow expressions like the following:
```
app(pay("alice", "bob", USD(100)));
```
only instead of submitting the transaction directly to the transaction engine it will submit the transaction through RPC and circulate it through a test network of peers / validators.

File: src/ripple/test/mao/impl/Net_test.cpp (hunk 3 above)
Comment (vinniefalco) [inline src/ripple/test/mao/impl/Net_test.cpp]: The fully qualified name of the test is `ripple.mao.Net`, where mao stands for "Multiple Application Objects."

File: src/ripple/test/jtx/impl/Env.cpp (hunk 2 above)
Comment (vinniefalco) [inline src/ripple/test/jtx/impl/Env.cpp]: The `Application` object is used in `Env::app()`. The `AppBundle` private class is necessary to allow construction in ctor-initializer lists.

File: src/ripple/test/mao/impl/Net_test.cpp (hunk 3 above)
Comment (ximinez) [inline src/ripple/test/mao/impl/Net_test.cpp]: Thanks.

**Comment by ximinez** [comment]
:+1:

File: src/ripple/test/mao/impl/Net_test.cpp (hunk 3 above)
Comment (vinniefalco) [inline src/ripple/test/mao/impl/Net_test.cpp]: We really should reverse the order of identifiers in the `BEAST_DEFINE_TESTSUITE` macro...

**Comment by scottschurr** [comment]
LGTM.  On OS X, both clang.debug.nounity and clang.release build with no warnings.  Unit tests and npm tests run successfully on both.  Also run/sync/stop on both.
:+1:

File: src/ripple/test/jtx/Env.h [hunk 4]
```diff
@@ -126,13 +127,21 @@ class Env
...
-
     /** The master account. */
     Account const master;
 
 private:
+    struct AppBundle
+    {
+        Application* app;
+        std::unique_ptr<Logs> logs;
+        std::unique_ptr<Application> owned;
```
Comment (ximinez) [inline src/ripple/test/jtx/Env.h]: Because only `app` is accessed anywhere (`Env::app()`), `logs` and `owned` could be made private.

Comment (vinniefalco) [inline src/ripple/test/jtx/Env.h]: The whole `struct` is already private, what would be the point of that

Comment (ximinez) [inline src/ripple/test/jtx/Env.h]: Oh, yeah, it is.

File: src/ripple/test/jtx/impl/Env.cpp (hunk 2 above)
Comment (ximinez) [inline src/ripple/test/jtx/impl/Env.cpp]: Sorry, I meant this particular ctor doesn't appear to be used (yet).

Comment (vinniefalco) [inline src/ripple/test/jtx/impl/Env.cpp]: True. I might remove it later

**Pull request closed without merging**